packages/next-swc/crates/next-api/src/dynamic_imports.rs

```diff
@@ -0,0 +1,395 @@
+    // dynamic()
+    let mut visitor = DynamicImportVisitor::new();
+    program.visit_with(&mut visitor);
+
+    if visitor.import_sources.is_empty() {
+        return Ok(OptionDynamicImportsMap::none());
+    }
+
+    let mut import_sources = vec![];
+    for import in visitor.import_sources.drain(..) {
+        // Using the given `Module` which is the origin of the dynamic import, trying to
+        // resolve the module that is being imported.
+        let dynamic_imported_resolved_module = *esm_resolve(
+            Vc::upcast(PlainResolveOrigin::new(
+                ecmascript_asset.await?.asset_context,
+                module.ident().path(),
+            )),
+            Request::parse(Value::new(Pattern::Constant(import.to_string()))),
+            Value::new(EcmaScriptModulesReferenceSubType::Undefined),
+            OptionIssueSource::none(),
+            IssueSeverity::Error.cell(),
+        )
+        .first_module()
+        .await?;
+
+        if let Some(dynamic_imported_resolved_module) = dynamic_imported_resolved_module {
+            import_sources.push((import, dynamic_imported_resolved_module));
+        }
+    }
+
+    Ok(Vc::cell(Some(Vc::cell((module, import_sources)))))
+}
+
+/// A visitor to check if there's import to `next/dynamic`, then collecting the
+/// import wrapped with dynamic() via CollectImportSourceVisitor.
+struct DynamicImportVisitor {
+    dynamic_ident: Option<Ident>,
+    pub import_sources: Vec<String>,
+}
+
+impl DynamicImportVisitor {
+    fn new() -> Self {
+        Self {
+            import_sources: vec![],
+            dynamic_ident: None,
+        }
+    }
+}
+
+impl Visit for DynamicImportVisitor {
+    fn visit_import_decl(&mut self, decl: &turbopack_binding::swc::core::ecma::ast::ImportDecl) {
+        // find import decl from next/dynamic, i.e import dynamic from 'next/dynamic'
+        if decl.src.value == *"next/dynamic" {
+            if let Some(specifier) = decl.specifiers.first().and_then(|s| s.as_default()) {
+                self.dynamic_ident = Some(specifier.local.clone());
+            }
+        }
+    }
+
+    fn visit_call_expr(&mut self, call_expr: &CallExpr) {
+        // Collect imports if the import call is wrapped in the call dynamic()
+        if let Callee::Expr(ident) = &call_expr.callee {
+            if let Expr::Ident(ident) = &**ident {
+                if let Some(dynamic_ident) = &self.dynamic_ident {
+                    if ident.sym == *dynamic_ident.sym {
+                        let mut collect_import_source_visitor = CollectImportSourceVisitor::new();
+                        call_expr.visit_children_with(&mut collect_import_source_visitor);
+
+                        if let Some(import_source) = collect_import_source_visitor.import_source {
+                            self.import_sources.push(import_source);
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
Don't we need to track scope declarations with clashing names? Would it just be easier to emit a magic comment as part of the dynamic transform that we could read at the module scope, like server actions does?

Reply:
> Don't we need to track scope declarations with clashing names?

Do we? any examples?

Reply:
```js
import dynamic from 'next/dynamic';

function Foo(dynamic) {
  dynamic(…)
}
```

Reply:
I would expect existing next_dynamic transform already conflicts with those edge case, since it injects `loadableGenerated` property against to the `dynamic` call if identifier clahses like this example. 

so in this example it'll be like

```
function Foo(dynamic) {
 dynamic(..., loadableGenerated: {}) // this'll cause error already
}
```

in most cases linter would complain even before though.